presenter/process.go

```diff
@@ -0,0 +1,167 @@
+	Links         map[string]Link            `json:"links"`
+}
+
+type ProcessResponseHealthCheck struct {
+	Type string                         `json:"type"`
+	Data ProcessResponseHealthCheckData `json:"data"`
+}
+
+type ProcessResponseHealthCheckData struct {
+	Type              string `json:"-"`
+	Timeout           int64  `json:"timeout"`
+	InvocationTimeout int64  `json:"invocation_timeout"`
+	HTTPEndpoint      string `json:"endpoint"`
+}
+
+func (h ProcessResponseHealthCheckData) MarshalJSON() ([]byte, error) {
+	timeout := &(h.Timeout)
+	if *timeout == 0 {
+		timeout = nil
+	}
+	invocationTimeout := &(h.InvocationTimeout)
+	if *invocationTimeout == 0 {
+		invocationTimeout = nil
+	}
+
+	switch h.Type {
+	case "http":
+		return json.Marshal(ProcessResponseHTTPHealthCheckData{
+			Timeout:           timeout,
+			InvocationTimeout: invocationTimeout,
+			HTTPEndpoint:      h.HTTPEndpoint,
+		})
+	case "port":
+		return json.Marshal(ProcessResponsePortHealthCheckData{
+			Timeout:           timeout,
+			InvocationTimeout: invocationTimeout,
+		})
+	case "process":
+		return json.Marshal(ProcessResponseProcessHealthCheckData{
+			Timeout: timeout,
+		})
+	default:
+		return json.Marshal(h)
+	}
+}
+
+type ProcessResponseHTTPHealthCheckData struct {
+	Timeout           *int64 `json:"timeout"`
+	InvocationTimeout *int64 `json:"invocation_timeout"`
+	HTTPEndpoint      string `json:"endpoint"`
+}
+
+type ProcessResponsePortHealthCheckData struct {
+	Timeout           *int64 `json:"timeout"`
+	InvocationTimeout *int64 `json:"invocation_timeout"`
+}
+
+type ProcessResponseProcessHealthCheckData struct {
+	Timeout *int64 `json:"timeout"`
+}
+
+type ProcessListResponse struct {
+	PaginationData PaginationData    `json:"pagination"`
+	Resources      []ProcessResponse `json:"resources"`
+}
+
+func ForProcess(responseProcess repositories.ProcessRecord, baseURL url.URL) ProcessResponse {
+	return ProcessResponse{
+		GUID:        responseProcess.GUID,
+		Type:        responseProcess.Type,
+		Command:     responseProcess.Command,
+		Instances:   responseProcess.Instances,
+		MemoryMB:    responseProcess.MemoryMB,
+		DiskQuotaMB: responseProcess.DiskQuotaMB,
+		HealthCheck: ProcessResponseHealthCheck{
+			Type: string(responseProcess.HealthCheck.Type),
+			Data: ProcessResponseHealthCheckData{
+				Type:              string(responseProcess.HealthCheck.Type),
+				Timeout:           responseProcess.HealthCheck.Data.TimeoutSeconds,
+				InvocationTimeout: responseProcess.HealthCheck.Data.InvocationTimeoutSeconds,
+				HTTPEndpoint:      responseProcess.HealthCheck.Data.HTTPEndpoint,
+			},
+		},
+		Relationships: map[string]Relationship{
+			"app": {
+				Data: RelationshipData{
+					GUID: responseProcess.AppGUID,
+				},
+			},
+		},
+		Metadata: Metadata{
+			Labels:      responseProcess.Labels,
+			Annotations: responseProcess.Annotations,
+		},
+		CreatedAt: responseProcess.CreatedAt,
+		UpdatedAt: responseProcess.UpdatedAt,
+		Links: map[string]Link{
```

Comment:
Do the links always have the same shape? If the API response will always have the same set of links, it may be worthwhile to make a `Links` struct to avoid magic strings. I agree that `map[string]string` is useful when it is unknown how many entries will be present (e.g.: labels).

Reply:
I favor the approach of making structs for the links since it gives a bit better inspection. If the concern is that we will have an empty `Link` because it isn't present, it can be omitted from the json response with the json libraries tags.

I'd like to hear from others on the team, though, since there is a large stylistic element to this. I could easily see a solution becoming a bit heavy handed. An example could be `Links` structs that have getter/stringer methods that could be embedded (since everything would get promoted).

@matt-royal - am I just asking for more trouble?

Reply:
Converted the map to use a struct for now, we can discuss here or in Team Forum later